### PR TITLE
Redux introduction

### PR DESCRIPTION
**Pull request by jankrzempek**
**What was implemented:** Redux toolkit based on the [https://redux-toolkit.js.org/tutorials/typescript](url) and square share.
**GOAL:** Store different app states using this tool

Main file that stores all the reducers is `store/index.ts`.
`thresholdSlice.ts` is the initial version for future slider and its values

--------------------------------------------------------------------------
Example of implementation:

Import:
```
import { useAppSelector, useAppDispatch } from '../../store/hooks'
import { updateGivenIndex } from '../../store/thresholdSlice'
```

Use:
```
 const threshold = useAppSelector((state) => (state.threshold));
 const dispatch = useAppDispatch()
```

Execute:
```
console.log("DID TAP ME")
dispatch(updateGivenIndex({id: 2, value: 2}))
```

### CODE REVIEW

File: app/javascript/react/App.tsx [hunk 1]
```diff
@@ -1,14 +1,18 @@
...
 import GlobalStyles from "./assets/styles/global-styles";
+import store from "./store/index";
+import './locales/i18n'
 
 const App = () => {
   return (
     <>
-      <GlobalStyles />
-      <CalendarPage />
+      <Provider store={store}>
```
Comment (franiakat) [inline app/javascript/react/App.tsx:12]: An empty tag is not needed anymore, the rule of returning a single child element is et thanks to adding the Provider component.

File: app/javascript/react/store/hooks.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import { useDispatch, useSelector } from 'react-redux'
+import type { TypedUseSelectorHook } from 'react-redux'
+import type { RootState, AppDispatch } from './index'
+
+export const useAppDispatch: () => AppDispatch = useDispatch
+export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector
```
Comment (franiakat) [inline app/javascript/react/store/hooks.ts:6]: We need to agree on prettier styling standards :D
```suggestion
import { useDispatch, useSelector } from "react-redux";
import type { TypedUseSelectorHook } from "react-redux";

import type { RootState, AppDispatch } from "./index";

export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
```

File: app/javascript/react/store/index.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import { configureStore } from "@reduxjs/toolkit";
+import thresholdReducer from "./thresholdSlice";
```
Comment (franiakat) [inline app/javascript/react/store/index.ts:2]: Please separate the imports: external libs and internal module.

File: app/javascript/react/store/thresholdSlice.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
+import { createSlice } from "@reduxjs/toolkit";
+import type { PayloadAction } from "@reduxjs/toolkit";
+import type { RootState } from "./index";
```
Comment (franiakat) [inline app/javascript/react/store/thresholdSlice.ts:3]: Imports rules. ;)

File: app/javascript/react/store/thresholdSlice.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  name: "threshold",
+  initialState,
+  reducers: {
+    updateAll: (state, action: PayloadAction<ThresholdState>) => {
+      state.th1 = action.payload.th1;
+      state.th2 = action.payload.th2;
+      state.th3 = action.payload.th3;
+      state.th4 = action.payload.th4;
+      state.th5 = action.payload.th5;
+    },
```
Comment (franiakat) [inline app/javascript/react/store/thresholdSlice.ts:36]: We can de-structure action payload object earlier and refer to particular `th` values when updating the state.
```suggestion
    updateAll: (
      state,
      { payload: { th1, th2, th3, th4, th5 } }: PayloadAction<ThresholdState>
    ) => {
      state.th1 = th1;
      state.th2 = th2;
      state.th3 = th3;
      state.th4 = th4;
      state.th5 = th5;
    },
```

File: app/javascript/react/store/thresholdSlice.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    updateGivenIndex: (state, action: PayloadAction<Threshold>) => {
+      const { id, value } = action.payload;
+      const key = `th${id}` as keyof ThresholdState;
+      console.log(key)
+      if (key in state) {
+        state[key] = value;
+      } else {
+        console.log("No update possible, the key is out of the range.");
+      }
+    },
```
Comment (franiakat) [inline app/javascript/react/store/thresholdSlice.ts:46]: I think I would change the Threshold interface as follows:
```
interface Threshold {
  name: keyof ThresholdState;
  value: number;
}
```

And then we could have this action defined:
```
    updateGivenIndex: (
      state,
      { payload: { name, value } }: PayloadAction<Threshold>
    ) => {
      state[name] = value;
    },
```

**Review by franiakat — CHANGES_REQUESTED** [review]
Thanks a lot for a great description, I have suggested a few changes. :)

I think we may end up with thresholds being a different data structure, but let's see once we get into working on API and data structure more.

File: app/javascript/react/store/thresholdSlice.ts (hunk 6 above)
Comment (jankrzempek) [reply app/javascript/react/store/thresholdSlice.ts:46]: Nice, it did not came through my mind, thanks!

**Review by franiakat — APPROVED** [review]
Good job, Janek, thanks for applying suggestions, ready to go! 🤸‍♂️

**Pull request merged**